File: scripts/importar_productos.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-
import json
import mysql.connector
from datetime import datetime, timedelta

fecha_actual = datetime.now().strftime("%Y-%m-%d %H:%M")
fecha = datetime.now().strftime("%Y%m%d")
path = 'tmp/productos'+fecha+'.json'

# Establecer la conexión con la base de datos
conexion = mysql.connector.connect(
    host="localhost",
    user="precios",
    password="precios",
    database="precios"
)

# Realizar operaciones en la base de datos
# Por ejemplo, ejecutar una consulta SQL
cursor = conexion.cursor()

productos_existentes = []
productos_erroneos = []
precios_existentes = []
registros_agregados = []
precios_cambiados = []
fecha_ultimo_precio_upd = []
NO_CAT = 155

def agregar_precio(cursor, reg):
    url_prod = ""
    if "url" in reg:
        url_prod = reg['url']
    text_nuevo_precio = "Se agregó nuevo precio - Carga Masiva - "+reg['name']+" - "+str(reg['price'])
    cursor.execute("INSERT INTO price (product_id, price, date_time, branch_id, es_oferta, confiabilidad, url) VALUES (%s, %s, %s, %s, %s, %s, %s)", (id_producto, reg['price'], fecha_actual, reg['branch_id'], 0, 100, url_prod))
    cursor.execute("INSERT INTO news (text, datetime, type_id) VALUES (%s, %s, %s)", (text_nuevo_precio, fecha_actual, 1))
          
with open(path) as archivo_json:
    precios = json.load(archivo_json)

for reg in precios:
    print(reg)
    #Buscamos en la base de datos en la tabla products si existe un producto con el mismo nombre
    cursor.execute("SELECT * FROM products WHERE name =  %s", (str(reg['name']),))
    resultados = cursor.fetchall()

    if not "category" in reg:
        reg['category'] = NO_CAT

    if reg['category'] == -1 or reg['branch_id'] == -1:
        print("Producto con categoria o local erroneo")
        productos_erroneos.append(reg)
        continue

    if len(resultados) == 0:
        cursor.execute("INSERT INTO products (name, vendor_id, ultimo_precio_conocido) VALUES (%s, %s, %s)", (reg['name'], reg['vendor_id'], fecha_actual))
        id_producto_agregado = cursor.lastrowid
        cursor.execute("INSERT INTO product_category (product_id, category_id) VALUES (%s, %s)", (id_producto_agregado, reg['category']))
        
        text_nuevo_precio = "Se agregó nuevo precio - Carga Masiva - "+reg['name']+" - "+str(reg['price'])
        url_prod = ""
        if "url" in reg:
            url_prod = reg['url']
        cursor.execute("INSERT INTO price (product_id, price, date_time, branch_id, es_oferta, confiabilidad, url) VALUES (%s, %s, %s, %s, %s, %s, %s)", (id_producto_agregado, reg['price'], fecha_actual, reg['branch_id'], 0, 100, url_prod))
        cursor.execute("INSERT INTO news (text, datetime, type_id) VALUES (%s, %s, %s)", (text_nuevo_precio, fecha_actual, 1))
        print("Se inserto el producto")
        registros_agregados.append(reg)
        
    else:
        print("El producto ya existe")
        id_producto = resultados[0][0]
        
        cursor.execute("UPDATE products SET ultimo_precio_conocido = %s WHERE id = %s", (fecha_actual, id_producto))
        cursor.execute("UPDATE products SET last_price = %s WHERE id = %s", (reg['price'], id_producto))
        cursor.execute("SELECT * FROM price WHERE product_id = %s and branch_id = %s AND confiabilidad > 90 ORDER BY date_time DESC LIMIT 1  ", (id_producto, reg['branch_id']))
        precio_db = cursor.fetchone()

        if (precio_db == None):
            print('No se encontro precio, se agrega')
            agregar_precio(cursor, reg)
            registros_agregados.append(reg)
        else:
            id_precio = precio_db[0]
            if abs(round(precio_db[2]) - round(reg['price'])) > 1:
                precios_cambiados.append([ reg,  precio_db ])
                agregar_precio(cursor, reg)
                text_nuevo_precio = "Se actualiza precio - Carga Masiva - "+reg['name']+" - "+str(reg['price'])
                cursor.execute("INSERT INTO news (text, datetime, type_id) VALUES (%s, %s, %s)", (text_nuevo_precio, fecha_actual, 1))
            else:
                url_prod = ""
                if "url" in reg:
                    url_prod = reg['url']
                cursor.execute("UPDATE price SET date_time = %s, url = %s WHERE id = %s", (fecha_actual, url_prod, id_precio))
                print('Ya existe un precio para ese articulo en ese local, se actualiza la fecha')
                #text_nuevo_precio = "Se reafirma precio - Carga Masiva - "+reg['name']+" - "+str(reg['price'])
                #cursor.execute("INSERT INTO news (text, datetime, type_id) VALUES (%s, %s, %s)", (text_nuevo_precio, fecha_actual, 1))
                fecha_ultimo_precio_upd.append(reg)
            precios_existentes.append(reg)

        productos_existentes.append(reg)

sql = "DELETE FROM news WHERE id NOT IN ( SELECT id  FROM ( SELECT id FROM news ORDER BY datetime DESC LIMIT 1000 ) AS subquery)"
#cursor.execute(sql)

sql = "DELETE FROM price WHERE price = 0"
cursor.execute(sql)

sql = "SELECT count(*) as total FROM price"
cantidad_total_precios = cursor.fetchall()

conexion.commit()

for precio in precios_cambiados:
    print(precio[0])
    print(precio[1])
    print("")

print("Cantidad total de precios: ", cantidad_total_precios)
print("Cantidad productos existentes: ", len(productos_existentes))
print("Cantidad total de precios: ", len(precios))
print("Cantidad registros erroneos: ", len(productos_erroneos))
print("Cantidad registros de precios existentes en local: ", len(precios_existentes))
print("Cantidad registros de precios agregados: ", len(registros_agregados))
print("Cantidad de precios cambiados: ", len(precios_cambiados))
print("Cantidad de precios en los que se actualizó su fecha de ultimo precio conocido: ", len(fecha_ultimo_precio_upd))
# Cerrar la conexión con la base de datos
conexion.close()

File: scripts/importar_productos.js
require("dotenv").config({ path: '../.env' })
const { text } = require("express");
const fs = require('fs');
const { resolve } = require("path");
const uuid = require("uuid")

let conn_obj = {
    host: process.env.mysql_host,
  //  port: process.env.mysql_port,
    user: process.env.mysql_user,
    password: process.env.mysql_password,
    database: process.env.mysql_database,
    supportBigNumbers: true,
    bigNumberStrings: true,
    typeCast: function (field, next) {
        if (field.type == "NEWDECIMAL") {
            var value = field.string();
            return (value === null) ? null : Number(value);
        }
        return next();
    }
  
  }
  
const knex = require('knex')({
    client: 'mysql2',
    connection: conn_obj,
    pool: { min: 0, max: 7 }
})


let precios_reafirmados = []
let precios_actualizados = []
let nuevos_precios_creados = []
let diccio_enterprise = {}
let diccio_branch = {}

async function nuevo_reg_precio( trx, articulo, producto_db, fecha_registro ){

    const insert = {
        "id": uuid.v7(),
        "product_id": producto_db.id,
        "price": articulo.price,
        "date_time": new Date(fecha_registro),
        "branch_id": articulo.branch_id,
        "es_oferta": 0,
        "confiabilidad": 100,
        "notas": (articulo?.nota) ? articulo.nota : null,
        "url": ( articulo.url ) ? articulo.url : null,
        "time": new Date(),
    }
    let insert_1 = await trx('price').insert( insert )
    precio_hoy = {...insert}
    precio_hoy['id'] = uuid.v4()
    let insert_2 = await trx('price_today').insert( precio_hoy )
    if (insert_1 && insert_2){
        nuevos_precios_creados.push( insert )
        return insert
    } else {
        return false
    }
}


exports.procesar_articulo = procesar_articulo

async function get_categoria( trx, articulo ){
    return new Promise( async (resolve, reject) => {
        let categoria = await knex('category').select().where('name', articulo.category_name).first()
        if (categoria){
            resolve(categoria)
            return
        } else {
            const ID_NUEVA_CAT = uuid.v7()
            const insert = { 'id': ID_NUEVA_CAT, 'name': articulo.category_name }
            categoria = await trx('category').insert( insert )
            resolve(insert)
            return
        }
    })
}

async function get_producto( trx, articulo ){
    return new Promise( async (resolve, reject) => {
        let producto  = await knex('products').select().where('name', articulo.name).first()
        if (producto){
            if (articulo?.barcode){
                await trx('products').update( {
                    "barcode": articulo.barcode
                } ).where('id','=',producto.id)
            }
            resolve(producto)
            return
        } else {
            const ID_NUEVO_PROD = uuid.v7()
            let insert = {
                "id": ID_NUEVO_PROD,
                "name": articulo.name,
                "vendor_id": articulo.vendor_id,
            }
            if (articulo?.barcode) insert['barcode'] = articulo.barcode
            let nuevo_reg = await trx('products').insert( insert )
            resolve(insert)
            return
        }
    })
}

async function procesa_precio( trx, producto_db, articulo, fecha_registro ){
    return new Promise( async (resolve, reject) => {
        try {
            let HOY = new Date()
            HOY.setHours(0,0,0,1)

            let ultimo_precio = await knex('price').select().where('product_id', producto_db.id).orderBy('time', 'desc').first()
            if (ultimo_precio){
                if (Math.abs(ultimo_precio.price - articulo?.price) > 1){
                    let nuevo_precio = await nuevo_reg_precio( trx, articulo, producto_db, fecha_registro )
                    if (nuevo_precio){
                        await trx('news').insert( {
                            text: "Se actualiza precio de "+articulo.name+" que ahora sale "+articulo.price
                        } )
                        await procesar_variacion(trx, [ ultimo_precio, articulo ], fecha_registro)
                        precios_actualizados.push( [ ultimo_precio, articulo ] )
                        return resolve(true)
                    } else 
                        return resolve(false)
                } else if (ultimo_precio && Math.abs(ultimo_precio.price - articulo?.price) <= 1){
                    await trx('price').update( {
                        "date_time": new Date(fecha_registro), "time": new Date(), "url": ( articulo.url ) ? articulo.url : null
                    } ).where("id", ultimo_precio.id)
                    let precio_hoy = {
                        ...ultimo_precio,
                        "date_time": new Date(fecha_registro), "time": new Date(), "url": ( articulo.url ) ? articulo.url : null
                    }
                    precio_hoy['id'] = uuid.v4()
                    //await trx('price_today').insert( precio_hoy )
                    precios_reafirmados.push(ultimo_precio)
                    return resolve(true)
                } else  if (!ultimo_precio) {
                    let nuevo_precio = await nuevo_reg_precio( trx, articulo, producto_db, fecha_registro )
                    if (nuevo_precio){
                        //await trx('news').insert( {
                        //    text: "Se agrega nuevo precio de "+articulo.name+" que sale "+articulo.price
                        //} )
                        return resolve(true)
                    } else 
                        return resolve(false)
                } else 
                    return resolve(false)
            } else {
                let nuevo_precio = await nuevo_reg_precio( trx, articulo, producto_db, fecha_registro )
                if (nuevo_precio){
                // await trx('news').insert( {
                //     text: "Se agrega nuevo precio de "+articulo.name+" que sale "+articulo.price
                //  } )
                    return resolve(true)
                } else 
                    return resolve(false)
            }

        } catch( error ){
            console.log(error)
            return resolve({ stat: false, text: error})
        }
    })
}

async function procesar_articulo(trx, articulo, fecha_registro ){
    return new Promise( async (resolve, reject) => {
        try {
            if (!articulo?.category_name){
                return resolve({stat:false, text: 'No se especifica categoria!'})
            }

            let res = { stat: true, text: '' }
            let proms_arr = []

            let producto  = await get_producto( trx, articulo )
            let categoria = await get_categoria( trx, articulo )
            
            //Si no hay relacion entre producto y categoria se la crea
            if (producto && categoria){
                let hay_cat = await knex('product_category').select()
                    .where({ "product_id": producto.id, "category_id": categoria.id }).first()
                if (!hay_cat)
                    proms_arr.push(
                        trx('product_category').insert( {
                            "id": uuid.v7(),
                            "product_id":  producto.id,
                            "category_id": categoria.id
                        } ))

                proms_arr.push( 
                    procesa_precio( trx, producto, articulo, fecha_registro )
                )

                let res_proms = await Promise.all( proms_arr )
                if (res_proms){
                    console.log(res_proms)
                    resolve(res)
                    return
                } else {
                    res.stat = false
                    resolve(res)
                    return
                }
            } else {
                return resolve({ stat: false, text: "no hay producto y/o categoria"})
            }
                
        } catch( error ){
            console.log(error)
            return resolve({ stat: false, text: error})
        }
    })
}

async function procesar_variacion( trx, variacion, fecha_registro){
    
    const reg_anterior = variacion[0]
    const reg_nuevo    = variacion[1] 
    console.log(157, variacion)
    const porcentage = (reg_nuevo.price - reg_anterior.price) / ( reg_anterior.price / 100 )
    if (porcentage > 50 || porcentage < -50){
        console.log(variacion)
    }
    if (reg_anterior.price != reg_nuevo.price && reg_nuevo.branch_id == reg_anterior.branch_id 
        && reg_anterior.price != 0 && reg_nuevo.price != 0){
        await trx("estadistica_aumento_diario").insert(
            {
                "id_producto":        reg_anterior.product_id,
                "branch_id":          reg_anterior.branch_id,
                "porcentaje_aumento": porcentage,
                "precio_ayer":        reg_anterior.price,
                "precio_hoy":         reg_nuevo.price,
                "nombre_producto":    reg_nuevo.name,
                "nombre_comercio":     diccio_enterprise[ diccio_branch[reg_nuevo.branch_id].enterprise_id ].name,
                "fecha_utlimo_precio": fecha_registro
            })
    }
    return
}
exports.procesar_variacion = procesar_variacion

async function generar_estadisticas_variacion_diaria(trx, variaciones){
    let proms = []
    await trx("estadistica_aumento_diario").delete()
    for (let i=0; i < variaciones.length; i++){
        proms.push( procesar_variacion(trx, variaciones[i]) )
    }
    await Promise.all( proms )
}

let diccio = {}

setTimeout( async ()=>{

    //const archivo_importacion = await fs.promises.readFile(rutaPrecios, 'utf8')
    //const array_importacion   = JSON.parse(archivo_importacion)

    let enterprises = await knex("enterprice").select()
    let branchs     = await knex("branch").select()

    if (/*array_importacion && */ enterprises && branchs){
        const proms_procesar = []
        let trx = await knex.transaction()
        /*for (let i=0; i < array_importacion.length; i++){
            const articulo = array_importacion[i]
            const reg_concat = articulo.name+articulo.price+articulo.branch_id
            if (diccio[reg_concat] == undefined){
                diccio[reg_concat] = 1
            } else 
                continue
            
            console.log("procesando", articulo)
            proms_procesar.push(procesar_articulo(trx, articulo))
        }*/
        for (let i =0; i < enterprises.length; i++)
            diccio_enterprise[enterprises[i].id ] = enterprises[i]
        for (let i=0; i < branchs.length; i++)
            diccio_branch[branchs[i].id ] = branchs[i]

        /*let procesada = await Promise.all(proms_procesar)
        if (procesada){
            await trx("news").delete().where("datetime", "<", HOY)
            await trx.commit()
            console.log("Cantidad de productos agregados: ", nuevos_productos.length)
            console.log("Cantidad de precios agregados: ", nuevos_precios_creados.length)
            console.log("Cantidad de precios reafirmados: ", precios_reafirmados.length)
            console.log("Cantidad de precios actualizados: ", precios_actualizados.length)
        }*/
    }
}, 100)

